fdbserver/TLogServer.actor.cpp

```diff
@@ -18,38 +18,71 @@
  * limitations under the License.
  */
 
-#include "flow/Hash3.h"
-#include "flow/UnitTest.h"
+#include <chrono>
+#include <type_traits>
+#include <vector>
+
+#include "fdbclient/FDBTypes.h"
+#include "fdbclient/KeyRangeMap.h"
 #include "fdbclient/NativeAPI.actor.h"
 #include "fdbclient/Notified.h"
-#include "fdbclient/KeyRangeMap.h"
 #include "fdbclient/RunTransaction.actor.h"
 #include "fdbclient/SystemData.h"
-#include "fdbclient/FDBTypes.h"
-#include "fdbserver/WorkerInterface.actor.h"
-#include "fdbserver/LogProtocolMessage.h"
-#include "fdbserver/TLogInterface.h"
-#include "fdbserver/Knobs.h"
-#include "fdbserver/IKeyValueStore.h"
-#include "fdbserver/MutationTracking.h"
-#include "flow/ActorCollection.h"
 #include "fdbrpc/FailureMonitor.h"
-#include "fdbserver/IDiskQueue.h"
+#include "fdbrpc/Stats.h"
 #include "fdbrpc/sim_validation.h"
 #include "fdbrpc/simulator.h"
-#include "fdbrpc/Stats.h"
-#include "fdbserver/ServerDBInfo.h"
+#include "fdbserver/FDBExecHelper.actor.h"
+#include "fdbserver/IDiskQueue.h"
+#include "fdbserver/IKeyValueStore.h"
+#include "fdbserver/Knobs.h"
+#include "fdbserver/LogProtocolMessage.h"
 #include "fdbserver/LogSystem.h"
-#include "fdbserver/WaitFailure.h"
+#include "fdbserver/MutationTracking.h"
+#include "fdbserver/PartMerger.h"
 #include "fdbserver/RecoveryState.h"
-#include "fdbserver/FDBExecHelper.actor.h"
+#include "fdbserver/ServerDBInfo.h"
+#include "fdbserver/TimedKVCache.h"
+#include "fdbserver/TLogInterface.h"
+#include "fdbserver/WaitFailure.h"
+#include "fdbserver/WorkerInterface.actor.h"
+#include "flow/ActorCollection.h"
+#include "flow/Hash3.h"
+#include "flow/UnitTest.h"
+
 #include "flow/actorcompiler.h"  // This must be the last #include.
 
 using std::pair;
 using std::make_pair;
 using std::min;
 using std::max;
 
+class TLogCommitRequestMerger : public PartMerger<UID, TLogCommitRequest> {
+protected:
+	virtual void merge(value_t& value, const value_t& incomingRequest) override {
+		ASSERT(value.prevVersion == incomingRequest.prevVersion);
+		ASSERT(value.version == incomingRequest.version);
+
+		// FIXME -- think about this
+		// ASSERT(value.knownCommittedVersion == incomingRequest.knownCommittedVersion);
+		// ASSERT(value.minKnownCommittedVersion == incomingRequest.minKnownCommittedVersion);
+		value.knownCommittedVersion = std::max(value.knownCommittedVersion, incomingRequest.knownCommittedVersion);
+		value.minKnownCommittedVersion =
+		    std::max(value.minKnownCommittedVersion, incomingRequest.minKnownCommittedVersion);
+
+		if (!value.additionalMessages.present()) {
+			value.additionalMessages = std::vector<std::pair<Arena, StringRef>>();
+		}
+		value.additionalMessages.get().emplace_back(std::make_pair(incomingRequest.arena, incomingRequest.messages));
```

Comment:
I am not sure if this is correct. This is basically saying adding `messages` in their receiving order. For split transactions, let's the buffer is `B`, and is divided into `B1`, `B2`, `B3`. If you merge them back, the order should be preserved, also probably strip any headers you added for `B1`, `B2`, `B3`. I suppose `merge` should construct `B` back?

Reply:
Actually this is an interesting thing -- does the order important? Intuitively I feel maybe no since we are basically saying for a given version V there are (K1, V1), (K2, V2), etc. But is this true? I am not quite sure. I have not added headers to the Arenas and corresponding (K, V) pairs.

Reply:
You are mostly right. However, I think it's legal to have two mutations i a transaction, which does: `a = 1`, `a = 3`. If they are split into 2 pieces, e.g., `B1` and `B2` in the example, then the order is important.